### ISSUE TITLE
Folding with higher folding factors than 2

### ISSUE THREAD

**Issue by Al-Kindi-0** [0xPolygonMiden/miden-vm#567]
As mentioned in #564, a higher folding factor can lead to a reduction in the number of cycles if combined with special instructions to offload all or most of the heavy parts during FRI verification. This issue aims at:

1. Finding the most suitable formulas to do folding for factors 4 and 8 with regards to Miden assembly.
2. Based on the above, decide on the parameter configuration that allows us to achieve the biggest savings for the smallest number, if any, of new instructions.
3. The above points should take into account the procedure for computing the folded position in the exponent from the value in the previous layer. Integrating this procedure into a potential folding instruction(s) might make sense.

**Comment by Al-Kindi-0** [comment]
### Introduction
During the FRI commit phase, the prover (in the interactive version) of the protocol computes a sequence of polynomials in response to a sequence of random challenges provided by the verifier. After each computation done by the prover, the latter gives oracle access to the evaluation of the computed polynomial over a domain that is larger by a given factor (called the blow-up factor) than the degree of the polynomial.

More precisely, the protocol starts with the evaluations over a certain domain $D_0$ of a polynomial $P_0(X)$ that the prover claims to be of degree $deg(P_0) < \delta_0$. In each round of the commit phase, the prover does the following:

1. Provide oracle access to the evaluation of the polynomial $P_{i-1}(X)$ of degree $deg(P_{i-1})< \delta_{i-1}$ over an evaluation domain $D_{i-1}$ of size $\delta_{i-1}/\rho$ where $\rho$ is what is called the rate.
2. Receive a random challenge $\alpha$ from the verifier.
3. Decompose the previous polynomial $P_{i-1}(X)$ of degree $deg(P_{i-1})< \delta_{i-1}$ into the following decomposition $$P_{i-1}(X) =  X^0 P_{i-1}^{0}(X^{r_i}) + X^1 P_{i-1}^{1}(X^{r_i}) \cdots + X^{r_i -1}P_{i-1}^{r_i - 1}(X^{r_i})$$ where $r_i$ is the folding factor for the the current step and $deg(P_{i-1}^j) < \frac{\delta_{i-1}}{r_i} := \delta_{i}$.
4. Compute the following polynomial $$P_{i}(Y) =  \alpha^0 P_{i-1}^{0}(Y) + \alpha^1 P_{i-1}^{1}(Y) \cdots + \alpha^{r_i -1}P_{i-1}^{r_i - 1}(Y)$$ which is of degree $\delta_i$ given the previous point.

In fact the above steps are compressed, in practice, such that the prover computes, given $\alpha$, the evaluations of $P_i$ on $D_i$ from those of $P_{i-1}$ on $D_{i-1}$ directly.  
The evaluation domain $D_0$ is chosen to be a subgroup of the multiplicative subgroup of the base field. The subsequent domains are generated via the recursive relationship $$D_i :=  \\{ x^{r_i} : x\in D_{i-1} \\}$$

For each of the maps $\(\cdot\)^{r_i}: D_{i-1} \to D_{i}$ we can define its kernel which is a cyclic group generated by a primitive $r_i$-th root of unity, call it $\tau$.

This issue is concerned with a verifier who, given $P_{i}(y),P_{i-1}(x_0),\cdots,P_{i-1}(x_{r_i-1})$ would like to ensure that the prover performed the folding step correctly. The naive approach is straightforward, solve the linear system of $r_i$ equations in $r_i$ variables, compute the linear combination using the challenge $\alpha$ and then accept if the values agree. In what follows, we investigate better methods that use the inherent algebraic structure of the folded polynomials.

## Recursive approach

### The $fold_2$ function:
To make the explanation clearer, let's consider the case where the folding factor is fixed and is equal to $2$. In this case, the kernel is just $\\{1,\tau\\}$ where $\tau = -1$ modulo $M$ the size of the base field. In this case, we have the following system of equations:

$$
\begin{cases}
 P_{i-1}(x) &=  P_{i-1}^{0}(x^{2}) + x P_{i-1}^{1}(x^{2}) \\
 P_{i-1}(x\tau) &=  P_{i-1}^{0}(x^{2}) + x\tau P_{i-1}^{1}(x^{2})
 \end{cases}       
$$ 

 Thus we can write it also in the following way

$$
\begin{cases}
 P_{i-1}(x) &=  P_{i-1}^{0}(x^{2}) + x P_{i-1}^{1}(x^{2}) \\
 P_{i-1}(x\tau) &=  P_{i-1}^{0}(x^{2}) - x P_{i-1}^{1}(x^{2})
 \end{cases}       
$$ 

This gives 

$$
\begin{cases}
 P_{i-1}^{0}(x^{2}) &=  (P_{i-1}(x) + P_{i-1}(x\tau))/2 \\
 P_{i-1}^{1}(x^{2})  &=  (P_{i-1}(x) - P_{i-1}(x\tau))/2x
 \end{cases}       
$$ 

and our assertion becomes 

$$P_{i}(y) = \frac{1}{2} (P_{i-1}(x) + P_{i-1}(x\tau) + \frac{\alpha}{x} (P_{i-1}(x) - P_{i-1}(x\tau)))$$

The above is in fact nothing but the $2$-point iFFT of $(P_{i-1}(x), P_{i-1}(x\tau))$, evaluated at $\frac{\alpha}{x}$. We can define this combined operation as:

$$ fold_2(\frac{\alpha}{x}; P_{i-1}(x), P_{i-1}(x\tau)) := \frac{1}{2} (P_{i-1}(x) + P_{i-1}(x\tau) + \frac{\alpha}{x} (P_{i-1}(x) - P_{i-1}(x\tau)))$$

### Formula for $fold_4$ 

In the case of a folding factor equal to $4$, we have the following system of equations:

$$
\begin{cases}
 P_{i-1}(x) &=  P_{i-1}^{0}(x^{4}) + x P_{i-1}^{1}(x^{4}) + x^2 P_{i-1}^{2}(x^{4}) + x^3 P_{i-1}^{3}(x^{4}) \\
 P_{i-1}(x\tau) &=  P_{i-1}^{0}(x^{4}) + x\tau P_{i-1}^{1}(x^{4}) + x^2\tau^2 P_{i-1}^{2}(x^{4}) + x^3\tau^3 P_{i-1}^{3}(x^{4}) \\
P_{i-1}(x\tau^2) &=  P_{i-1}^{0}(x^{4}) + x\tau^2 P_{i-1}^{1}(x^{4}) + x^2\tau^4 P_{i-1}^{2}(x^{4}) + x^3\tau^6 P_{i-1}^{3}(x^{4}) \\
P_{i-1}(x\tau^3) &=  P_{i-1}^{0}(x^{4}) + x\tau^3 P_{i-1}^{1}(x^{4}) + x^2\tau^6 P_{i-1}^{2}(x^{4}) + x^3\tau^9 P_{i-1}^{3}(x^{4})
 \end{cases}       
$$ 

Again, this gives us that $P_i(y)$ is nothing but the iFFT of $(P_{i-1}(x), P_{i-1}(x\tau), P_{i-1}(x\tau^2), P_{i-1}(x\tau^3))$ evaluated at $\frac{\alpha}{x}$. Using the fact that the $4$-points iFFT can be constructed from $2\times 2$-points iFFT, we can combine the iFFT and the evaluation, using the $fold_2$ procedure to get:

$$
\begin{align*}
P_i(y) &:=  fold_4(\frac{\alpha}{x}; P_{i-1}(x), P_{i-1}(x\tau), P_{i-1}(x\tau^2), P_{i-1}(x\tau^3)) \\
 &=  fold_2(\frac{\alpha^2}{x^2}; v_0, v_1)
\end{align*}
$$

where

$$
\begin{align*}
v_0 &:= fold_2(\frac{\alpha}{x};P_{i-1}(x), P_{i-1}(x\tau^2)) \\
v_1 &:= fold_2(\frac{\alpha}{x}\tau;P_{i-1}(x\tau), P_{i-1}(x\tau^3))
\end{align*}
$$

Note  the bit-reversed ordering of $(P_{i-1}(x), P_{i-1}(x\tau), P_{i-1}(x\tau^2), P_{i-1}(x\tau^3))$.

### Formula for $fold_8$

Similarly we get, for $(u_0,\cdots,u_7)$ the bit-reversed ordering of $(P_{i-1}(x),\cdots,P_{i-1}(x\tau^7))$, that:

$$
\begin{align*}
P_i(y) &:=  fold_8(\frac{\alpha}{x}; P_{i-1}(x),\cdots, P_{i-1}(x\tau^7)) \\
 &=  fold_2(\frac{\alpha^4}{x^4}; v_0, v_1)
\end{align*}
$$

where 

$$
\begin{align*}
v_0 &:= fold_4(\frac{\alpha}{x};v_0,\cdots,v_3) \\
v_1 &:= fold_4(\frac{\alpha}{x}\tau;v_4,\cdots,v_7)
\end{align*}
$$

### Formula for $fold_{16}$

for $(u_0,\cdots,u_{15})$ the bit-reversed ordering of $(P_{i-1}(x),\cdots,P_{i-1}(x\tau^{15}))$, that:

$$
\begin{align*}
P_i(y) &:=  fold_{16}(\frac{\alpha}{x}; P_{i-1}(x),\cdots, P_{i-1}(x\tau^{16})) \\
 &=  fold_2(\frac{\alpha^8}{x^8}; v_0, v_1)
\end{align*}
$$

where 

$$
\begin{align*}
v_0 &:= fold_8(\frac{\alpha}{x};v_0,\cdots,v_7) \\
v_1 &:= fold_8(\frac{\alpha}{x}\tau;v_8,\cdots,v_{15})
\end{align*}
$$

## The direct approach

Note that the previous method combines in an efficient way the two operations of:

1. Interpolation over the set $(P_{i-1}(x),\cdots,P_{i-1}(x\tau^{r_i -1}))$.
2. Evaluation of the resulting polynomial at $\frac{\alpha}{x}$.

So the ultimate goal is to evaluate a polynomial at a point, given only its evaluation over a given set of points. The barycentric formula is the appropriate tool to achieve this goal.

More precisely, for a polynomial given by its evaluations  $(y_0,\cdots,y_{r -1}))$ over the interpolation set $(x,x\tau,\cdots,x\tau^{r-1})$, its evaluation at a point $\frac{\alpha}{x}$ outside of its interpolation set, is given by:
$$\frac{1}{r}\left(\left(\frac{\alpha}{x}\right)^r - 1 \right)\sum\limits_{i=0}^{r -1}\frac{y_j\tau^j}{\frac{\alpha}{x}-\tau^j}$$

or equivalently

$$\frac{1}{r}\left(\left(\frac{\alpha}{x}\right)^r - 1 \right)\sum\limits_{i=0}^{r -1}\frac{y_j}{\frac{\alpha}{x\tau^j}-1}$$

## Discussion of the costs

For the recursive approach, the approximate cost estimate is $$(r-1)*(2A_e + M_e) + (log_2(r) - 1)*M_e + (\frac{r}{2} - 1)*M$$ for a constant folding factor $r$ where

1. $A_e$ and $M_e$ are the costs of an addition and a multiplication in the extension field, respectively.
2. $M$ is the cost of a multiplication in the base field.

For the direct approach, assuming non-determinism is used for efficient inversion, we have the approximate cost of $$r*(2M + A + M_e) + (log_2(r) - 1)*M_e + M + A$$
where, in addition to the notation from before, $A$ is an addition in the base field. The companion Rust [code](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L437) illustrates how the above strategies are implemented.

## A new Miden VM instruction

In the case of Miden, the two above approaches are very close from a cost perspective, in theory. In practice, the recursive approach has the advantage that an instruction to do $fold_4$, for example, can be introduced in the VM and hence make it the most competitive approach.

In the case of extension degree $2$, the inputs to the $fold_4$ Miden instruction would be $4$ extension field elements i.e. 8 base field elements, and the evaluation point in the extension field $\frac{\alpha}{x}$ which is $2$ base field elements. This means that the inputs on the stack to the $fold_4$ instruction are going to be the top $10$ stack elements.
In the case of extension degree $3$, we have $5$ extension field elements using the same reasoning as above. This means that, in the case of extension degree $3$, the top $15$ elements are going to be the input to the $fold_4$ instruction. 

We can do still better. Assume that the current position index is $p$ and current domain size is $d$ and $poe = g^{p}$ where g is the current domain generator (assuming the offset is the identity element). Let $(q,res) := (p/(d/r), p \\% (d/r))$ where the division is integer division.
In the [current](https://github.com/maticnetwork/miden/blob/3ad1533d3dae232762cecbb536756d49d806d21d/stdlib/asm/crypto/fri/frif2e2.masm#L123) implementation we see that the `verify_query_layer` procedure does essentially two things:

1. Get the Merkle leaf for the folded position and unhash it to get the values $y_0,...,y_{r-1}$ in the extension field. 
2. Check that the previous evaluation $e_{old}$ is equal to $y_i$ where $i = q$.
3. Compute the new evaluation $e_{new}$ using $y_0,...,y_r$, $\alpha$, $poe$ and $res$. $res$ is used to compute $x$, which is used to compute the evaluation point for the folding procedure, $\frac{\alpha}{x}$, as $x = poe/\tau^{res}$

This means that we can have an instruction that takes $(y_0,...,y_{r-1}, res, e_{old}, poe, \alpha)$ and produces $(e_{new}, poe^r)$ so as to include the computation done by the `next_pos_exp` into the folding instruction.

Thus in Miden assembly we would have, in the case of folding factor $2$, something like

```
# Input:  [d_size, p, C, t_depth, e1, e0, poe, a1, a0, ...]
# Output: [t_depth-1, e1_new, e0_new, p%d_size/2, d_size/2, poe^2, ...]
export.verify_query_layer.1
    exec.check_hash         #[v3, v2, v1, v0, ?, e1_old, e0_old, poe, a1, a0, p%d_size/2, d_size/2, t_depth,...]
    exec.fold_2             #[e1_new, e0_new, poe^2, p%d_size/2, d_size/2, t_depth,...]
    exec.post_ops           #[t_depth-1, e1_new, e0_new, p%d_size/2, d_size/2, poe^2, ...]
end
```
and in the case of folding factor equal to $4$

```
# Input:  [d_size, p, C, t_depth, e1_old, e0_old, poe, a1, a0, ...]
# Output: [t_depth-1, e1_new, e0_new, p%d_size/4, d_size/4, poe^4, ...]
export.verify_query_layer.1
    exec.check_hash         #[v7, v6, v5, v4, v3, v2, v1, v0, ?, e1_old, e0_old, poe, a1, a0, p%d_size/4, d_size/4, t_depth,...]
    exec.fold_4             #[e1_new, e0_new, poe^4, p%d_size/4, d_size/4, t_depth,...]
    exec.post_ops           #[t_depth-1, e1_new, e0_new, p%d_size/4, d_size/4, poe^4, ...]
end
```
where the comment to the right of the executed procedure represents its resulting stack configuration and where `check_hash` and `post_ops` are two procedures introduced in order to improve readability.
Thus the $fold_4$ VM instruction will have as input the first $14$ elements of the stack and will output $3$ elements at the top of the stack, in the case of extension degree 2.

**Comment by bobbinth** [comment]
This is awesome! And the code is super helpful.

As a next step, I think we need to understand a bit deeper how we will implement `fold4` procedure in a single VM instruction. I'm specifically thinking about whether we can reduce the degree of the instruction to only 2.

Looking at how we perform folding [here](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L446), I see that we have the following operations:

1. `let evaluation_point = alpha.mul_base(x.inv())` - here the degree of `evaluation_point` would be 2.
2. `evaluation_point.mul_base(tau.inv())` - here, the degree of the resulting expression would be 2 because `tau` is a constant (correct?).
3. `evaluation_point * evaluation_point` - here the degree of resulting expression would be 4.

Thus, we'd exceed our desired degree 2 as soon as we use any of these in an expression. A solution would be to use helper registers (we have 6 of them available) to put results of these expressions there. For example:

- Helper registers `h0` and `h1` would hold the result of `alpha.mul_base(x.inv())`.
- Helper registers `h2` and `h3` would hold the result of `evaluation_point * evaluation_point`.
- It doesn't seem like we need to do anything for `evaluation_point.mul_base(tau.inv())` because we can values in `h0` and `h1` to compute it on they fly without increasing the degree.

Next, `fold2` procedure requires a multiplication and thus also raises the degree of the result. For example, if all inputs are of degree 1, the output would be of degree 2. Thus, we again need to apply degree reduction technique to values `tepm0` and `temp1` so that they are of degree 1 by the time we call last `fold2`. This would require 4 more registers to hold intermediate values - we can use some of stack registers for that.

Lastly, we need to compute inverse of `x`. We can do this non-deterministically, but this also requires one helper register. We can use `h4` for this.

Transition diagram just for this operation (not including other things like `poe` computation) could look like this:

![image](https://user-images.githubusercontent.com/43513081/206485885-e4d2acfc-7b25-4b85-a928-6073c72a52fe.png)

Where:
- `d0 ... d7` encode 4 points in the original domain.
- `a0, a1` is the $\alpha$.
- `ev0, ev1` is the result of `alpha.mul_base(x.inv())`.
- `es0, es1` is the result of `evaluation_point * evaluation_point` computed using values in `ev0` and `ev1`.
- `t0, t1` encodes the result off the first invocation of `fold2` (basically, the value of [tmp0](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L452)).
- `s0, s1` encodes the result off the second invocation of `fold2` (basically, the value of [tmp1](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L457)).
- `en0, e1` encodes a point in the folded domain.

Does this sound about right?

**Comment by Al-Kindi-0** [comment]
> ...because tau is a constant (correct?).

That is indeed the case.

> Does this sound about right?

Yes, it makes a lot of sense and I don't see any flaws in the reasoning above. Do you think that there is a possibility of including the `poe` related computations as well?

**Comment by bobbinth** [comment]
> Do you think that there is a possibility of including the `poe` related computations as well?

I think so. This was just my attempt to start with a single thing. Now we can add `poe` to this.

**Comment by bobbinth** [comment]
The next thing I'm thinking about is how to get `x` onto the stack. The formula for computing `x` is [here](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L437). I think we should be able to roll this into the `FRIFOLD4` operation.

First, we need to compute `current_pos / trarget_domain_size` (and we actually need to perform similar computations [here](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L397) and [here](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L430)). Since both values are guaranteed to be 32-bit integers, we can use `u32unchecked_divmod` instruction:
```
# stack: [t_domain_size, current_pos]
u32unchecked_divmod
# stack: [f_pos, d_seg]
```
Where `d_seg` is the "domain segment" computed as `current_pos / target_domain_size`. Its value can be 0, 1, 2, or 3. `f_pos` is the value of the folded position.

The next question is how to get the required power of `tau` onto the stack. We can do that using `exp.u2` instruction, that that'll probably take 10 - 12 cycles. Another approach is allow the prover to provide binary decomposition of `d_seg` value non-deterministically, then then use these values in the `FRIFOLD4` operation.

Let's call first and second bits of `d_seg` `ds0 and `ds1` respectively. Then, a full `FRIFOLD4` operation could look like so:

![image](https://user-images.githubusercontent.com/43513081/206885124-07817d1c-6d19-4fec-aa0c-0a9dbcef9b74.png)

Most of the things above are either self-explanatory or have been covered previously in this thread, but a couple deserve additional explanation.

- The reason why we compute `poe^2` is to reduce degree for `poe^4` computation. Basically, we use one of the registers in the next state of the stack to hold the intermediate value of `poe^2` and thus, reduce the degree of related transition constraints to 2.
- As mentioned above, `ds0` and `ds1` is the binary decomposition of `d_seg` provided by the prover non-deterministically. The constraints we'll need to impose to make sure this is correct are:
  - Both `ds0` and `ds1` are binary.
  - `ds0 + 2 * ds1 = d_seg`
- We can use `ds0` and `ds1` to compute the required power of `tau` as follows: `f_tau = (1-ds0) * (1-ds1) + tau * ds0 * (1-ds1) + tau^2 * (1-ds0) * ds1 + tau^3 * ds0 * ds1`. Since `tau` is a constraint for this computation would be of degree 2.
- We can use `f_tau` to compute `x` as `x = pow / f_tau`

After `FRIFOLD4` operation is done, we don't really care about values one the stack beyond `poe^4`. So, these could be treated as "garbage" by the subsequent operation.

**Comment by Al-Kindi-0** [comment]
This is great! Also, the idea of using the remaining "output slots" to reduce the degree is cool.

**Comment by bobbinth** [comment]
I think the last piece left here is figuring out how to handle checking of previous evaluation efficiently (i.e., [these lines](https://github.com/Al-Kindi-0/miden/blob/2a53d5df9d6a8fa79e945c1ee56b3ebd352ad2a0/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L423-L435)). We have all query values on the stack (`v0, ..., v7`), but selecting the right one based on the value of `d_seg` will take quite a few cycles. Thus, it would be good to do this check as a part of `FRIFOLD4` as well.

One issue is that in my previous proposal we don't have the value of previous evaluation on the stack. Another issue is that to compute selector flags for `d_seg` we need to multiply `ds0` and `ds1` together, and this is already a degree 2 constraint (and thus, we can't use it to select non-constant values). Here is another proposal which should address these.

Stack transition for this new proposal could look like so:

![image](https://user-images.githubusercontent.com/43513081/211189070-40ae5610-4f9c-4138-b79b-e271650d73d7.png)

In the above `pe0` and pe1` encode the value of previous evaluation. Also, we decompose `d_seg` into 4 mutually exclusive binary flags: `ds0, ..., ds3`. These flags are set as follows:
- `ds0 = 1` when `d_seg = 0`.
- `ds1 = 1` when `d_seg = 1`.
- `ds2 = 1` when `d_seg = 2`.
- `ds3 = 1` when `d_seg = 3`.

Otherwise, these flags equal to 0.

Then, to check that previous evaluation is equal to the correct query value, we can enforce the following constraints:
```
(v0 - pe0) * ds0 + (v2 - pe0) * ds1 + (v4 - pe0) * ds2 + (v6 - pe0) * ds3 = 0
(v1 - pe1) * ds0 + (v3 - pe1) * ds1 + (v5 - pe1) * ds2 + (v7 - pe1) * ds3 = 0
```

We'll also need to re-write how `f_tau` is computed. But the new formula is even simpler:
```
f_tau = ds0 + tau * ds1 + tau^2 * ds2 + tau^3 * ds3
```
And enforcing that `d_seg` and `ds` values are consistent, would need to be done like so:
```
d_seg = (ds0 + ds1 * 2 + ds2 * 3 + ds3 * 4) - 1
```

One last thing to note: I removed `d_size` and `t_depth` from the stack. As far as I can tell, they are not required for the `FRIFOLD4` operation (but we do need them earlier). My thinking here is that we can pre-compute them for each layer and store them next to $\alpha$ values. Specifically, for each layer, we'd store a word in memory which could look something like this: `[d_size, t_depth, a0, a1]`. Then, when we start processing a new layer, we can read this word from memory and we'll have all the values that we need.

@Al-Kindi-0 - am I missing anything here?

**Comment by Al-Kindi-0** [comment]
Looks great to me! 
The introduction of the exclusive flag indicating the segment simplifies things a lot.

> One last thing to note: I removed d_size and t_depth from the stack. As far as I can tell, they are not required for the FRIFOLD4 operation (but we do need them earlier). My thinking here is that we can pre-compute them for each layer and store them next to values. Specifically, for each layer, we'd store a word in memory which could look something like this: [d_size, t_depth, a0, a1]. Then, when we start processing a new layer, we can read this word from memory and we'll have all the values that we need.

It seems to me that even with the costs of precomputations and memory management required to compute `d_size` and `t_depth`, comparing evaluations as part of `FRIFOLD4` will be cheaper. So I think this solution makes sense.

**Comment by Al-Kindi-0** [comment]
Now that we have a good idea of how `FRIFOLD4` should look like, it would be great to decide on the best configuration of its input and output stack-states. To do so, we can summarize the context around `FRIFOLD4` in the following incomplete pseudo-code:

```
# Input: [num_iter, pos, poe, pe1, pe0, ...] where pos is the query, poe is offset * initial_domain_generator^pos,
# pe1 and pe0 are the evaluations.
# Output: [...] the final evaluations i.e. remainder[f_pos] where f_pos is pos % remainder_size
proc.verify_query
    push.1
    while.true
        exec.verify_query_layer
        dup
        neq.0
    end
    exec.verify_remainder
end


# Input: [num_iter, pos, poe, pe1, pe0, ...]
# Output: [num_iter - 1, f_pos, poe^4, ne1, ne0, ...]
proc.verify_query_layer

    # Load layer-specific data [d_size, t_depth, a1, a0]
    # [d_size, t_depth, a1, a0, num_iter,  pos, poe, pe1, pe0, num_iter, ...]
    dup
    movdn.5
    push.0.0.0.0
    dup.4
    # TODO: use num_iter to compute correct memory pointer
    loc_loadw.x

    # Get leaf node
    # [V', f_pos, d_seg, a1, a0, poe, pe1, pe0, num_iter,  ...]
    movup.5
    swap
    u32unchecked_divmod
    push.0.0.0.0
    movup.9
    #TODO: use num_iter to compute correct memory pointer
    loc_loadw.x                
    dup.4 
    movup.7               
    mtree_get                   
    swap 
    drop                   

    # TODO: Unhash leaf
    # [v7, v6, v5, v4, v3, v2, v1, v0, f_pos, d_seg, a1, a0, poe, pe1, pe0, num_iter,  ...]

    # TODO: Apply FRIFOLD4 and some post-processing
    # [num_iter - 1, f_pos, poe^4, ne1, ne0, ...]
end
```
`verify_remainder` fetches the `remainder[f_pos]` and asserts that it is equal to `(ne0, ne1)`.
It seems that it might make sense to include the number of iterations in `FRIFOLD4` as there is still one free slot available, is this accurate @bobbinth ? There is also the possibility that there is a cleaner way to do the above with better reference management.

**Comment by bobbinth** [comment]
Looks great! Thank you! A few comments:

I'm imagining the memory to be laid out like this:

![image](https://user-images.githubusercontent.com/43513081/212461245-7d443fb8-f06e-4eca-bb8e-c5313f1760a6.png)

Where `r0, ..., r2` represent a root of a Merkle tree for a given layer.

In this case, instead of passing `num_iter` we could just pass `layer_ref` which would be pointer to the $n$-th layer. And then we just need to increment the pointer by $2$ in every iteration.

Next, I don't think we can use `loc_loadw` instructions here because for these instructions the index must be specified statically. Instead, we can use `mem_loadw` which could read the layer pointer off the top of the stack.

Lastly, it might be possible to get rid of at least one `push.0.0.0.`. We have a lot of slots with "garbage" values after `FRIFOLD4` and maybe we can just overwrite these values when loading from memory. Figuring out how to do this cleanly might be more complicated than it is worth though.

**Comment by Al-Kindi-0** [comment]
In case we do away with `num_iter`, how do we keep track of the layer we are currently at?
One solution is to test whether `poe^4` is the 64th primitive root of unity, but maybe there's a better way.

**Comment by bobbinth** [comment]
> In case we do away with `num_iter`, how do we keep track of the layer we are currently at?

We need this only in `verify_query` for loop exit condition, right? If so, we could have another variable there something like `last_layer_ref` and then instead of doing `neq.0` we'd compare `last_layer_ref` with the `layer_ref` currently on the top of the stack. This will probably add 5 extra cycles or so to each iteration of the loop, but it might be more efficient than trying to compute memory offsets based on `num_iter`.

**Comment by Al-Kindi-0** [comment]
To summarize the discussion so far, this is what we have:
1. Input stack + helper registers:
`[v7, v6, v5, v4, v3, v2, v1, v0, a1, a0, poe, f_pos, d_seg, pe1, pe0, ...] + [ev0, ev1, es0, es1, x, x_inv]`

2. Output stack + helper registers:
`[en1, en0, f_pos, poe^4, poe^2, t1, t0, s1, s0, ds3, ds2, ds1, ds0, f_tau, ...] + [...]`

3. Folding formulas:
    1. `fold2(a, b, e) := ((a + b) + (a - b) * e)/2` or `(r0, r1) == fold2(a, b, e) if and only if (2*r0, 2*r1) = a + b + e* (a - b)`

    2. `(t0, t1) := fold2(qv0, qv2, ev)` where `qv0 := (v0, v1); qv2 := (v4, v5); ev:= (ev0, ev1)`

    3. `(s0, s1) := fold2(qv1, qv3, ev * tau_inv)` where `qv1 := (v2, v3); qv3 := (v6, v7); es:= (es0, es1)`
    4. `(en0, en1) := fold2((t0, t1), (s0, s1), (es0, es1))`

4. Eval point
    1. `alpha := (a0, a1)`
    2. `(ev0, ev1) := alpha * x_inv`
    3. `(es0, es1) := (ev0, ev1) * (ev0, ev1)`

5. `x`:
    1. `x = (poe/f_tau) * domain_offset`
    2. `f_tau = ds0 + tau * ds1 + tau^2 * ds2 + tau^3 * ds3`
    3. `ds0 = 1` if and only if `d_seg = 0`
    4. `ds1 = 1` if and only if `d_seg = 1`
    5. `ds2 = 1` if and only if `d_seg = 2`
    6. `ds3 = 1` if and only if `d_seg = 3`

6. correct d_seg decompostion:
    * `d_seg = (ds0 + 2 * ds1 + 3 * ds2 + 4 * ds3) - 1`

7. Check previous evaluation:
    1. `(v0 - pe0) * ds0 + (v2 - pe0) * ds1 + (v4 - pe0) * ds2 + (v6 - pe0) * ds3 = 0`
    1. `(v1 - pe1) * ds0 + (v3 - pe1) * ds1 + (v5 - pe1) * ds2 + (v7 - pe1) * ds3 = 0`

8. `poe^4` as `poe^2 * poe^2`:
    * `poe^4 = poe^2 * poe^2`

**Comment by bobbinth** [comment]
Closed by #706.
